tempesta_fw/sched/tfw_sched_ratio.c

```diff
@@ -0,0 +1,467 @@
+
+/*
+ * The size of pool of TfwRatioRated{} entries for RCU substitution.
+ * It's best when it's one more than the real number of CPU cores.
+ */
+#define TFW_RATIO_EPOOLSZ	(16 + 1)
+
+/*
+ * The size of scale that is used to range back-end servers.
+ * It's most effective in calculations when it's a power of two. 
+ */
+#define TFW_RATIO_SCALE_64	(1 << 6)
+#if TFW_SG_MAX_SRV > TFW_RATIO_SCALE_64
+#define TFW_RATIO_SCALE		TFW_SG_MAX_SRV
+#else
+#define TFW_RATIO_SCALE		TFW_RATIO_SCALE_64
+#endif
+
+/**
+ * List of connections to an upstream server.
+ * Connections can go up and down during failover process. Only
+ * fully established connections are considered by the scheduler.
+ */
+typedef struct {
+	atomic64_t		counter;
+	size_t			conn_n;
+	TfwServer		*srv;
+	TfwSrvConn 		*conns[TFW_SRV_MAX_CONN];
+} TfwRatioSrv;
+
+/*
+ * RCU-fortified array of pointers to TfwRatioSrv{} ranged by each
+ * back-end server's weight. The number of times a server is found
+ * in the array is defined by the server's weight relative to other
+ * back-end servers in the group.
+ */
+typedef struct {
+	struct rcu_head		rcu;
+	size_t			listsz;
+	TfwRatioSrv		*list[TFW_RATIO_SCALE];
+	atomic_t		free;
+} TfwRatioRated;
+
+/**
+ * List of upstream servers.
+ * The list is considered static, i.e. all servers, either dead
+ * or alive, are present in the list during the whole run-time.
+ * That may change in the future.
+ */
+typedef struct {
+	atomic64_t		counter;
+	size_t			srv_n;
+	atomic_t		rearm;
+	struct timer_list	timer;
+	TfwRatioSrv		srvs[TFW_SG_MAX_SRV];
+	TfwRatioRated		epool[TFW_RATIO_EPOOLSZ];
+	TfwRatioRated __rcu	*rated;
+} TfwRatioSrvList;
+
+static void
+tfw_sched_ratio_calc(int *weight, int *ratio, size_t sz)
+{
+	size_t s;
+	unsigned long calc[sz], sum = 0, unit;
+
+	for (s = 0; s < sz; ++s) {
+		calc[s] = ((1 << 16) * TFW_RATIO_SCALE) / weight[s];
+		sum += calc[s];
+	}
+	unit = ((1 << 16) * TFW_RATIO_SCALE) / sum;
+	for (s = 0; s < sz; ++s)
+		ratio[s] = (calc[s] * unit) / (1 << 16) ? : 1;
+}
+
+static inline void
+tfw_sched_fill_rated(TfwSrvGroup *sg, int *ratio, TfwRatioRated *rated)
+{
+	size_t s, i, n;
+	TfwRatioSrvList *sl = sg->sched_data;
+
+	for (s = 0, i = 0; s < sl->srv_n; ++s)
+		for (n = 0; n < ratio[s]; ++n, ++i)
+			rated->list[i] = &sl->srvs[s];
+	rated->listsz = i;
+}
+
+static inline void
+tfw_sched_fill_equal(TfwSrvGroup *sg, int *ratio, TfwRatioRated *rated)
+{
+	size_t s, i, n;
+	int eqshare = ratio[0];
+	TfwRatioSrvList *sl = sg->sched_data;
+
+	for (n = 0, i = 0; n < eqshare; ++n)
+		for (s = 0; s < sl->srv_n; ++s, ++i)
+			rated->list[i] = &sl->srvs[s];
+	rated->listsz = i;
+}
+
+/*
+ * Calculate server ratios based on specific dynamic weights.
+ * Return 1 if no recalculation is required.
+ * Return 0 if the ratios were recalculated.
+ */
+static int
+tfw_sched_ratio_dynamic(TfwSrvGroup *sg, int *ratio)
+{
+	size_t s;
+	TfwRatioSrvList *sl = sg->sched_data;
+	int recalc = 0, weight[sl->srv_n];
```

Comment:
With #76 in mind `sl->srv_n` can be very large, so we can easily overrun the stack.